in_toto/verifylib.go

```diff
@@ -0,0 +1,517 @@
+              // successfully consumed, i.e. it will be removed from the queue
+              consumed = append(consumed, srcPath)
+            }
+            queue = Subtract(queue, consumed)
+
+          case "allow":
+            consumed := FnFilter(ruleData["pattern"], queue)
+            queue = Subtract(queue, consumed)
+
+          case "disallow":
+            disallowed := FnFilter(ruleData["pattern"], queue)
+            if len(disallowed) > 0 {
+              return fmt.Errorf("Artifact verification failed for %s '%s'." +
+                  " Artifact(s) %s disallowed by rule %s.",
+                  reflect.TypeOf(itemI), itemName, disallowed, rule)
+            }
+
+          // TODO: Support create, modify, delete rules
+          default:
+              return fmt.Errorf("Cannot process artifact rule '%s'. We" +
+                  " don't support rules of type '%s'", rule,
+                  strings.ToUpper(ruleData["type"]))
+        }
+      }
+    }
+  }
+  return nil
+}
+
+
+func ReduceStepsMetadata(layout Layout,
+    stepsMetadata map[string]map[string]Metablock) (map[string]Metablock,
+    error) {
+  stepsMetadataReduced := make(map[string]Metablock)
+
+  for _, step := range layout.Steps {
+    linksPerStep, ok := stepsMetadata[step.Name]
+    // We should never get here, layout verification must fail earlier
+    if !ok || len(linksPerStep) < 1 {
+      panic("Could not reduce metadata for step '" + step.Name +
+          "', no link metadata found.")
+    }
+
+    // Get the first link (could be any link) for the current step, which will
+    // serve as reference link for below comparisons
+    var referenceKeyID string
+    var referenceLinkMb Metablock
+    for keyID, linkMb := range linksPerStep {
+      referenceLinkMb = linkMb
+      referenceKeyID = keyID
+      break
+    }
+
+    // Only one link, nothing to reduce, take the reference link
+    if len(linksPerStep) == 1 {
+      stepsMetadataReduced[step.Name] = referenceLinkMb
+
+    // Multiple links, reduce but first check
+    } else {
+      // Artifact maps must be equal for each type among all links
+      // TODO: What should we do if there are more links, than the
+      // threshold requires, but not all of them are equal? Right now we would
+      // also error.
+      for keyID, linkMb := range linksPerStep {
+        if !reflect.DeepEqual(linkMb.Signed.(Link).Materials,
+            referenceLinkMb.Signed.(Link).Materials) ||
+            !reflect.DeepEqual(linkMb.Signed.(Link).Products,
+            referenceLinkMb.Signed.(Link).Products) {
+          return nil, fmt.Errorf("Link '%s' and '%s' have different" +
+              " artifacts.",
+              fmt.Sprintf(LinkNameFormat, step.Name, referenceKeyID),
+              fmt.Sprintf(LinkNameFormat, step.Name, keyID))
+        }
+      }
+      // We haven't errored out, so we can reduce
+      stepsMetadataReduced[step.Name] = referenceLinkMb
+    }
+  }
+  return stepsMetadataReduced, nil
+}
```

Comment:
I somehow think this is a threshold signature check. Should we name it like so?

Reply:
Rather not, there is a separate `VerifyLinkSignatureThesholds` function that gets called earlier. We could merge `VerifyLinkSignatureThesholds` and `ReduceStepsMetadata`, but then we can't verify command alignment for all links anymore, which we do after verifying signature thresholds, but before merging all links for a step.